### PR TITLE
Dynamic lib loading

### PR DESCRIPTION
**Pull request by kurtuluso**
Platform independent dynamic loading feature is required for loading DDS Security plugins. It is also required for unit tests of the plugins.

In this pull request, the following functions are added to dds_rt:
**ddsrt_dlopen**: loads the dynamic shared object
**ddsrt_dlclose**: informs the system that the library, identified by 'handle', is no longer needed
**ddsrt_dlsym**: gets the memory address of a symbol from a loaded library
**ddsrt_dlerror**: gets the most recent library related error

The functions are implemented for windows and posix platforms that support dynamic loading.

### CODE REVIEW

**Comment by martinbremmer** [comment]
Some comments:
1. It would have been nice if this PR only contained one or a few commits. Now it contains the difference between master and security branch as well.
1. Make ddsrt_dlopen() and ddsrt_dlsym() more consistent. Either let both return dds_retcode_t or let both return a pointer.
1. The handle param comment in the ddsrt_dlopen() header is not correct.
1. Add a small explanation to DDS_RETCODE_BAD_PARAM retval comment in the ddsrt_dlopen() header. Even if it seems a bit redundant.
1. Why not return DDS_RETCODE_BAD_PARAM in ddsrt_dlopen() when handle is NULL? Be aware that handle is set first thing. If it's NULL, it'll crash. Please add tests that checks what happens when either (or both) name and/or handle is NULL.
1. ddsrt_dlopen() checks for NULL pointer (at least for the name argument). All the other functions just assert. Please make it more consistent. Either let ddsrt_dlopen() assert with NULL as well, or add NULL checks to the other functions (and add tests to check NULL usage if you choose the latter).

**Comment by eboasson** [comment]
Thanks @kurtuluso for the pull request and @martinbremmer for reviewing it. I'd just like to point out that I am hopeful that the minor details left on #177 will be resolved soon and that it can be merged shortly. That would make it trivial to address the first point.

**Comment by kurtuluso** [comment]
1. Actually, there is only one commit but the security main branch is not up to date. And thıs wıll not be changed until be'ng prepared to merge to master
2.  DONE
3. DONE
4. DONE
5. DONE
6. DONE

**Comment by kurtuluso** [comment]
@eboasson 
This is ready for review (all pass in local travis: https://travis-ci.org/kurtuluso/cyclonedds/builds/536755903?utm_source=github_status&utm_medium=notification) but it still shows conflict.
I've rebased to updated security branch and readded my changes but still, it is reporting conflict in the pull request.
I think it needs your touch.

File: src/ddsrt/include/dds/ddsrt/library.h [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+ *
+ * The function ddsrt_dlclose() informs the system that the
+ * library, identified by 'handle', is no longer needed.
+ * will get the memory address of a symbol,
+ * identified by 'symbol', from a loaded library 'handle'.
+ *
+ * @param[in]   handle      Library handle.
+ *
+ * @returns A dds_retcode_t indicating success or failure.
+ *
```
Comment (martinbremmer) [inline src/ddsrt/include/dds/ddsrt/library.h]: The fact that ddsrt_dlclose() can return DDS_RETCODE_BAD_PARAMETER, should be mentioned.

File: src/ddsrt/src/library/posix/library.c [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    return retcode;
+}
+
+dds_retcode_t
+ddsrt_dlsym(
+    ddsrt_lib handle,
+    const char *symbol,
+    void **address)
+{
+    *address = NULL;
```
Comment (martinbremmer) [inline src/ddsrt/src/library/posix/library.c]: If address is NULL, this'll crash.

File: src/ddsrt/src/library/posix/library.c [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+dds_retcode_t
+ddsrt_dlerror(
+    char *buf,
+    size_t buflen)
+{
+  const char *err = dlerror();
+  if (err == NULL) {
+    return DDS_RETCODE_NOT_FOUND;
+  }
+  snprintf(buf, buflen, "%s", err);
```
Comment (martinbremmer) [inline src/ddsrt/src/library/posix/library.c]: If buf is NULL, this'll crash.

File: src/ddsrt/src/library/windows/library.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+dds_retcode_t
+ddsrt_dlerror(char *buf, size_t buflen)
+{
+  /* Hopefully (and likely), the last error is
+   * related to a Library action attempt. */
+  DWORD err = GetLastError();
+
+  /* Get the string that is related to this error. */
+  if (ddsrt_strerror_r(err, buf, buflen) == DDS_RETCODE_OK) {
```
Comment (martinbremmer) [inline src/ddsrt/src/library/windows/library.c]: This'll assert (debug) or likely crash (release) when buf is NULL.

File: src/ddsrt/tests/dl.c [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+  printf("\n%s", buffer);
+}
+
+CU_Test(ddsrt_library, dlopen_null)
+{
+  char buffer[256];
+  dds_retcode_t r;
+  ddsrt_lib l;
+
+  r = ddsrt_dlopen(NULL, false, &l);
```
Comment (martinbremmer) [inline src/ddsrt/tests/dl.c]: I also expect a test with the library handle being NULL.

File: src/ddsrt/tests/dl.c [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+  f_set(set_int);
+  get_int = f_get();
+  CU_ASSERT_EQUAL(set_int, get_int);
+
+  r = ddsrt_dlclose(l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+}
+
+CU_Test(ddsrt_library, dlclose_null)
```
Comment (martinbremmer) [inline src/ddsrt/tests/dl.c]: Why is this called dlclose_null? There's only one test with NULL. Please split this up in multiple tests.

File: src/ddsrt/tests/dl.c [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+
+    r = ddsrt_dlclose( l ); /*already closed handle */
+    CU_ASSERT_EQUAL(r, DDS_RETCODE_ERROR);
+
+    r = ddsrt_dlclose( NULL );
+    CU_ASSERT_EQUAL(r, DDS_RETCODE_BAD_PARAMETER);
+
+}
+
```
Comment (martinbremmer) [inline src/ddsrt/tests/dl.c]: Where are the ddsrt_dlerror() that supply NULL to their arguments?

File: src/ddsrt/include/dds/ddsrt/library.h [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+ * @param[in]   translate   Automatic name translation on/off.
+ * @param[out]  handle      Automatic name translation on/off.
+ *
+ * @returns A dds_retcode_t indicating success or failure.
+ *
+ * @retval DDS_RETCODE_OK
+ *             Library handle was successfully loaded.
+ * @retval DDS_RETCODE_BAD_PARAM
+ *             There is an invalid input in the parameter list
+ *
```
Comment (martinbremmer) [inline src/ddsrt/include/dds/ddsrt/library.h]: Remove this white-line.

File: src/ddsrt/include/dds/ddsrt/library.h [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+ * The function ddsrt_dlerror() will return the most recent library
+ * related error in human readable form.
+ *
+ * If no error was found, it's either due to the fact that there
+ * actually was no error since init or last ddsrt_dlerror() call,
+ * or due to an unknown unrelated error.
+ *
+ * @returns A dds_retcode_t indicating success or failure.
+ *
```
Comment (martinbremmer) [inline src/ddsrt/include/dds/ddsrt/library.h]: When NULL checks are added to ddsrt_dlerror(), then DDS_RETCODE_BAD_PARAMETER should be added to this header.

File: src/ddsrt/tests/dl.c [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+  r = ddsrt_dlsym(l, "get_int", &f);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL(f);
+  TEST_ABORT_IF_NULL(f, "ddsrt_dlsym(l, \"get_int\") failed.");
+
+  r = ddsrt_dlclose(l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+}
+
+CU_Test(ddsrt_library, dlsym_bad)
```
Comment (martinbremmer) [inline src/ddsrt/tests/dl.c]: Please rename this to dlsym_null

File: src/ddsrt/tests/dl.c [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+  r = ddsrt_dlopen(TEST_LIB_NAME, true, &l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL(l);
+  TEST_ABORT_IF_NULL(l,"ddsrt_dlopen() failed. Is the proper library path set?");
+
+  r = ddsrt_dlsym(l, NULL, &f);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_BAD_PARAMETER);
+
+  r = ddsrt_dlsym(NULL, "UnknownSym", &f);
```
Comment (martinbremmer) [inline src/ddsrt/tests/dl.c]: Please use a valid symbol name when you only want to check what happens the the library handle is NULL.

File: src/ddsrt/tests/dl.c [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_BAD_PARAMETER);
+
```
Comment (martinbremmer) [inline src/ddsrt/tests/dl.c]: I'd also expect a test that checks what happens when the symbol argument is NULL.

**Comment by martinbremmer** [comment]
I don't have any more remarks.

File: src/ddsrt/src/library/posix/library.c [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+      dlerror();    /* Clear any existing error */
+      if ( (translate) && (strrchr(name, '/') == NULL)) {
+        /* Add lib and suffix to the name and try to open. */
+#if __APPLE__
+        static const char suffix[] = ".dylib";
+#else
+        static const char suffix[] = ".so";
+#endif
+        char* libName = ddsrt_malloc(3 + strlen(name) + sizeof(suffix));
+        sprintf(libName, "lib%s%s", name, suffix);
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/posix/library.c]: Better not use `sprintf` here, but `snprintf`. If only to avoid compiler warnings. `ddsrt_asprintf` would be an option to, that way you can also ditch the `ddsrt_malloc`.

File: src/ddsrt/src/library/posix/library.c [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+  dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+  if ( name == NULL || handle == NULL ){
+      retcode = DDS_RETCODE_BAD_PARAMETER;
+
+  } else {
+      *handle = NULL;
+      dlerror();    /* Clear any existing error */
+      if ( (translate) && (strrchr(name, '/') == NULL)) {
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/posix/library.c]: I understand the `strrchr` is there to check if it's not a path, but that might not be a good enough check(?). It might just be that the user means a file in the current directory, so if there's no `/` in the name, perhaps do a `stat` as well? Maybe @eboasson has something on this?

File: src/ddsrt/src/library/posix/library.c [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+
+dds_retcode_t
+ddsrt_dlclose(
+    ddsrt_lib handle)
+{
+    dds_retcode_t retcode;
+
+    if( handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/posix/library.c]: Just use an early return here. Perhaps just settle on an `assert` or just don't do anything if `handle` is `NULL`.

File: src/ddsrt/src/library/posix/library.c [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    ddsrt_lib handle,
+    const char *symbol,
+    void **address)
+{
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+    if ( symbol == NULL || handle == NULL || address == NULL){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    } else {
+        *address =  dlsym(handle, symbol);
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/posix/library.c]: Use a local variable and only assign `address` on success.

File: src/ddsrt/src/library/posix/library.c [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+dds_retcode_t
+ddsrt_dlerror(
+    char *buf,
+    size_t buflen)
+{
+
+    const char *err;
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+    if( buf == NULL ){
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/posix/library.c]: An `assert` would be better in this case.

File: src/ddsrt/src/library/windows/library.c [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+    if ( name == NULL || handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+
+    } else {
+        *handle = NULL;
+        GetLastError(); /* Clear any existing error */
+        if ((translate) && (strrchr(name, '/') == NULL )
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/windows/library.c]: Same as with the POSIX comments.

File: src/ddsrt/src/library/windows/library.c [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+dds_retcode_t
+ddsrt_dlclose(
+    ddsrt_lib handle)
+{
+    dds_retcode_t retcode;
+
+    if( handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    }
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/windows/library.c]: Please choose a style and stick with it.

File: src/ddsrt/src/library/windows/library.c [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+            retcode = DDS_RETCODE_ERROR;
+        }
+    }
+
+    return retcode;
+}
+
+
+dds_retcode_t
+ddsrt_dlerror(char *buf, size_t buflen)
```
Comment (k0ekk0ek) [inline src/ddsrt/src/library/windows/library.c]: This doesn't do the same thing as the `ddsrt_dlerror` implementation in the POSIX layer above. `GetLastError` is used by much more of the api. It's better to use a thread-local variable to store the error when it occurred in one of the functions. Of course just dropping the `ddsrt_dlerror` call entirely is also a possibility. @eboasson do you have a preference here?

File: src/ddsrt/include/dds/ddsrt/library.h [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+
+#include "dds/export.h"
+#include "dds/ddsrt/types.h"
+#include "dds/ddsrt/retcode.h"
+
+#if defined (__cplusplus)
+extern "C" {
+#endif
+
+typedef void *ddsrt_lib;
```
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/ddsrt/include/dds/ddsrt/library.h]: I've added _t suffix but which _standard_ are we talking about? Is there a document to adhere? I do not remember now but I saw types without _t suffix.
Windows: 

```
typedef void *PVOID;
typedef PVOID HANDLE;
typedef HANDLE HINSTANCE;
typedef HINSTANCE HMODULE;
HMODULE LoadLibraryA(
  LPCSTR lpLibFileName
);

```

File: src/ddsrt/src/library/windows/library.c [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+{
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+    if ( name == NULL || handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+
+    } else {
+        *handle = NULL;
+        GetLastError(); /* Clear any existing error */
```
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/ddsrt/src/library/windows/library.c]: User should run the GetLastError if there is an error.
I've removed the line that has reset-purpose. As stated in the function explanation, the user should check error only if he/she gets DDS_RETCODE_ERROR.

File: src/ddsrt/src/library/windows/library.c [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+ddsrt_dlopen(
+    const char *name,
+    bool translate,
+    ddsrt_lib *handle)
+{
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+    if ( name == NULL || handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
```
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/ddsrt/src/library/windows/library.c]: early return is not a good practice, I try single return as much as I can. If you have a previous decision, I can change it.
I remember this practice in JOINT STRIKE FIGHTER AIR VEHICLE C++ CODING STANDARDS 

> AV Rule 113 (MISRA Rule 82, Revised)
Functions will have a single exit point.
Rationale: Numerous exit points tend to produce functions that are both difficult to
understand and analyze.
Exception: A single exit is not required if such a structure would obscure or otherwise
significantly complicate (such as the introduction of additional variables) a function’s control
logic. Note that the usual resource clean-up must be managed at all exit points.

File: src/ddsrt/src/library/posix/library.c (hunk 16 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/posix/library.c]: I think it is safer to reset the given argument to NULL if there is an error and the dlsym already sets it to NULL in case of error.

File: src/ddsrt/src/library/windows/library.c (hunk 18 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/windows/library.c]: done

File: src/ddsrt/src/library/windows/library.c (hunk 20 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/windows/library.c]: Yes, you're right. However, GetLastError stores the error message for each thread. So, if the user calls the ddsrt_dlerror just after calling the other library functions, he/she will get the error message properly.

File: src/ddsrt/src/library/windows/library.c (hunk 19 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/windows/library.c]: what do you mean?

File: src/ddsrt/src/library/posix/library.c (hunk 17 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/posix/library.c]: As we provide these functions to other components, we decided to convert asserts to NULL checks.
If you think that it is impossible to get buf == NULL, I can change it.

File: src/ddsrt/src/library/posix/library.c [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+ddsrt_dlopen(
+    const char *name,
+    bool translate,
+    ddsrt_lib *handle)
+{
+  dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+  if ( name == NULL || handle == NULL ){
+      retcode = DDS_RETCODE_BAD_PARAMETER;
```
[the remark this answers is not in the document]
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: no checking and ddsrt_nonnull_all would be even better

File: src/ddsrt/src/library/posix/library.c (hunk 14 above)
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: @k0ekk0ek this is actually the sensible thing to do: this way you can always avoid the translation by specifying ./file. I'd prefer to see "strchr" rather than "strrchr" though. (Also for windows.)

File: src/ddsrt/src/library/posix/library.c (hunk 13 above)
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: As this would be the _only_ use of ``sprintf`` ... yes, please replace it.

File: src/ddsrt/src/library/posix/library.c [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+
+  } else {
+      *handle = NULL;
```
[the remark this answers is not in the document]
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: agree

File: src/ddsrt/src/library/posix/library.c [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+#endif
+        char* libName = ddsrt_malloc(3 + strlen(name) + sizeof(suffix));
+        sprintf(libName, "lib%s%s", name, suffix);
+        *handle = dlopen(libName, RTLD_GLOBAL | RTLD_NOW);
+        ddsrt_free(libName);
+
+
+      }
+
+      if (*handle == NULL && retcode == DDS_RETCODE_OK) {
```
[the remark this answers is not in the document]
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: I do actually care about the layout in general and really dislike a bunch of blank lines before closing a block ... there are plenty of other details that I really prefer differently but those do irk me so much.

Checking ``*handle`` makes sense (that's simply so a successful dlopen on the "translated" name is handled correctly). The ``retcode`` check makes no sense here.

File: src/ddsrt/src/library/posix/library.c [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+        /* name contains a path,
+         * (auto)translate is disabled or
+         * dlopen on translated name failed. */
+        *handle = dlopen(name, RTLD_GLOBAL | RTLD_NOW);
```
[the remark this answers is not in the document]
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: if only the rest of the code were perfectly consistent on that point ... the documentation comment ought to indicate what the effect on the out parameter is in case of failure

File: src/ddsrt/src/library/posix/library.c (hunk 15 above)
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: agree

File: src/ddsrt/src/library/posix/library.c (hunk 16 above)
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: I'm have a bigger problem with the coding style (this is not the only place) of declaring-and-initialising a return value to "OK" and then changing it on error. *If* one insists on single return and initialising all variables at declaration time as "safety" measures, *then* the logical thing would be to initialise a return code to "ERROR".

File: src/ddsrt/src/library/posix/library.c (hunk 17 above)
Comment (eboasson) [reply src/ddsrt/src/library/posix/library.c]: I don't think @k0ekk0ek 's point is that it is impossible to get a null pointer for buf; it is rather that there is a good argument to make consider it illegal to do so, i.e., that a precondition is that buf != NULL. That means an assert, preferably a compiler annotation that it must not be a null pointer. The only callers will be internal and there seems to be no good reason to ever call it with a null pointer for buf. Checking for a null pointer will only hide errors.

File: src/ddsrt/src/library/windows/library.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (eboasson) [reply src/ddsrt/src/library/windows/library.c]: @kurtuluso : please read what you quoted:

> A single exit is not required if such a structure would obscure or otherwise
significantly complicate (such as the introduction of additional variables) a function’s control
logic. Note that the usual resource clean-up must be managed at all exit points.

An early return here reduces the amount of state on needs to track mentally (no null pointers — so no need to account for null pointers anywhere in the function anymore). No state has been changed yet, so an early return is the better option.

File: src/ddsrt/src/library/windows/library.c (hunk 19 above)
Comment (eboasson) [reply src/ddsrt/src/library/windows/library.c]: There are various possibilities — could be whitespace in combination with parenthesis ...

Also this is another of those functions where early returns and ternary expressions would be easier to read.

File: src/ddsrt/src/library/windows/library.c (hunk 20 above)
Comment (eboasson) [reply src/ddsrt/src/library/windows/library.c]: That's not what the documentation blurb says ("Get the most recent library related error.") You're now suggesting that it should be extended with "and you may not have intervening calls to any Win32 functions on the same thread". While I agree that commonly one would expect dlerror to be called immediately, there is something to be said in favour of implemting the interface properly.

File: src/ddsrt/tests/CMakeLists.txt [hunk 28]
```diff
@@ -10,9 +10,11 @@
 # SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 #
 include(CUnit)
+include(GenerateExportHeader)
 
 set(sources
     "atomics.c"
+    "dl.c"
```
Comment (eboasson) [inline src/ddsrt/tests/CMakeLists.txt]: just like "library" is a bit vague, "dl" is a bit vague, too

File: src/ddsrt/include/dds/ddsrt/library.h [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+ *             Symbol was found in the loaded library.
+ *             Address parameter is ready to use.
+ * @retval DDS_RETCODE_BAD_PARAM
+ *             There is an invalid input in the parameter list
+ * @retval DDS_RETCODE_ERROR
+ *             Symbol was not found.
+ *             Use ddsrt_dlerror() to diagnose the failure.
+ */
+DDS_EXPORT dds_retcode_t
+ddsrt_dlsym(
```
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/ddsrt/include/dds/ddsrt/library.h]: We've mapped posix and windows functions in ddsrt. Which posix function are you talking about?

File: src/ddsrt/src/library/posix/library.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/ddsrt/src/library/posix/library.c]: In DDS Security case, the parameters are coming from the configuration file and the user of this function probably get these parameters on the fly (not static), so I think it is better to check here.

File: src/ddsrt/src/library/posix/library.c (hunk 16 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/posix/library.c]: The initial result value is OK because, in this approach, there can be different error codes; the correct one will be assigned according to conditions. If there had only one error code and several valid (OK) return codes, I would have used ERROR initialisation at the beginning.

File: src/ddsrt/src/library/posix/library.c (hunk 27 above)
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/ddsrt/src/library/posix/library.c]: Parameter documentation has been updated:
> @param[out]  handle      Library handle that will be assigned after successfull operation. It is assigned to NULL if loading fails.

File: src/ddsrt/src/library/posix/library.c (hunk 15 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/posix/library.c]: @eboasson What is the parameter handling policy in Cyclone?
I will refactor all parameter handlings according to the policy.

File: src/ddsrt/include/dds/ddsrt/library.h (hunk 29 above)
[the remark this answers is not in the document]
Comment (eboasson) [reply src/ddsrt/include/dds/ddsrt/library.h]: dlsym — POSIX and Windows guarantee function pointers and data pointers are interchangeable, but C does not.

File: src/ddsrt/src/library/windows/library.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/ddsrt/src/library/windows/library.c]: It is ~20 lines of function in total. Using a single return here does not make it complicated or introduce new additional variables. Of course, early return always reduces the amount of state on needs to track mentally but I think it does not fit the exception.
Anyway, I will not continue the discussion anymore, I'm just adding a return.

File: src/ddsrt/src/library/windows/library.c (hunk 20 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/windows/library.c]: I've updated the description

File: src/ddsrt/tests/CMakeLists.txt (hunk 28 above)
Comment (kurtuluso) [reply src/ddsrt/tests/CMakeLists.txt]: dynlib

File: src/ddsrt/src/library/posix/library.c (hunk 17 above)
Comment (kurtuluso) [reply src/ddsrt/src/library/posix/library.c]: I've changed all parameter checks to assert. We need an assert policy in the coding guideline. (Of course, we need a guideline first)

**Comment by eboasson** [comment]
@kurtuluso there are still various things that I want to see changed, but I am getting the feeling that that would ultimately just end up in the absence of any pull requests for a security implementation. So, let's do it this way: I'll merge it into the security branch in the hope of seeing interesting things in the near future. Merging from the security branch into master is then going to be a painful affair again, but it will then (hopefully) be worth the pain.

Would you be so kind as to cherry-pick the CMake change I did for the Windows builds? There is Windows-specific code in here, so it has get through the CI build successfully before I can merge it.

Regarding your question on the use of assertions: the traditional guideline applies. Assertions are for checking invariants — and preconditions are invariants, too, so assertions are fine for checking parameters on internal functions. External interfaces — whether that be network, files, parameters in API functions — need to be validated with proper error handling, and so for those checks, assertions may not be used.

The ddsrt library is a platform abstraction for use by the Cyclone implementation, as well as any brave souls who decide to use it in their application. Misuse of ddsrt is therefore an affair punishable with exceptions and/or aborts raised by failed assertions.

The reason behind this is that one of the key tricks when writing software is to minimise the amount of state to be kept track of and to avoid having to handle cases that cannot in reality exist. Handling situations where a precondition was not met introduces more state than recognising that by definition such a situation may never occur, and therefore more complexity and a higher likelihood of further failures or mistakes.

At all times, static analysis is preferred over dynamic checks, so if the compiler or an easily accessible static analysis tool can guarantee the precondition, that is preferred over an assertion. Unforunately, Microsoft's SAL annotations are too platform specific to really be of use in this project; I still need to experiment with Frama-C.
